algorithms/src/Kokkos_RadixSort.hpp

```diff
@@ -0,0 +1,268 @@
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_RADIXSORT_HPP_
+#define KOKKOS_RADIXSORT_HPP_
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE_NOTDEFINED_RADIXSORT
+#endif
+
+#include <Kokkos_Core.hpp>
+
+namespace Kokkos {
+namespace Experimental {
+
+// https://developer.nvidia.com/gpugems/gpugems3/part-vi-gpu-computing/chapter-39-parallel-prefix-sum-scan-cuda
+
+template <int BitWidth, typename KeyView>
+struct KeyFromView {
+  using key_value_type          = typename KeyView::value_type;
+  static constexpr int num_bits = BitWidth > 0 ? BitWidth
+                                               : sizeof(key_value_type) * 8;
+
+  KeyView const& keys;
+  KeyFromView(KeyView const& k) : keys(k) {}
+
+  // i: index of the key to get
+  // bit: which bit, with 0 indicating the least-significant
+  auto operator()(int i, int bit) const {
+    auto h = keys(i) >> bit;
+
+    // Handle the sign bit of signed 2's-complement indicating low values
+    if constexpr (std::is_signed_v<key_value_type>) {
+      if (bit == 8 * sizeof(key_value_type) - 1) {
+        h = ~h;
+      }
+    }
+    return ~h & 0x1;
+  }
+
+  int getNumBits() { return num_bits; }
+};
+
+// I'd like to use CTAD for this, and skip the builder function, but
+// haven't figured out whether the syntax is possible for the user to
+// specify BitWidth, and the compiler to deduce KeyView
```

Comment:
Make the constructor take `std::integral_constant<int, BitWidth>` as 2nd argument